keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/multiple/LambdaModelVertexTest.java

```diff
@@ -0,0 +1,231 @@
+        inputToModel = inputToModelOne.plus(inputToModelTwo);
+        weatherModel.setInputToModel(inputToModel);
+
+        Map<VertexLabel, Vertex<? extends Tensor>> inputs = new HashMap<>();
+        inputs.put(new VertexLabel("Temperature"), inputToModel);
+
+        ModelVertex model = new LambdaModelVertex(inputs, this::modelExecution, this::updateValues);
+        DoubleVertex chanceOfRain = model.getDoubleModelOutputVertex(new VertexLabel("ChanceOfRain"));
+        DoubleVertex humidity = model.getDoubleModelOutputVertex(new VertexLabel("Humidity"));
+
+        DoubleVertex temperatureReadingOne = new GaussianVertex(chanceOfRain, 5);
+        DoubleVertex temperatureReadingTwo = new GaussianVertex(humidity, 5);
+        temperatureReadingOne.observe(3.0);
+        temperatureReadingTwo.observe(60.0);
+
+        NonGradientOptimizer gradientOptimizer = NonGradientOptimizer.of(temperatureReadingTwo.getConnectedGraph());
+        gradientOptimizer.maxLikelihood();
+        Assert.assertEquals(30.0, inputToModel.getValue().scalar(), 0.1);
+    }
+
+    @Test
+    public void modelWorksAsPartOfSampling() {
+        inputToModel = new GaussianVertex(25., 5);
+        weatherModel.setInputToModel(inputToModel);
+
+        Map<VertexLabel, Vertex<? extends Tensor>> inputs = new HashMap<>();
+        inputs.put(new VertexLabel("Temperature"), inputToModel);
+
+        ModelVertex model = new LambdaModelVertex(inputs, this::modelExecution, this::updateValues);
+        DoubleVertex chanceOfRain = model.getDoubleModelOutputVertex(new VertexLabel("ChanceOfRain"));
+        DoubleVertex humidity = model.getDoubleModelOutputVertex(new VertexLabel("Humidity"));
+
+        //My prior belief is the temperature is 29.0.
+        //These observations are indicative of a temperature of 30.
+        DoubleVertex chanceOfRainObservation = new GaussianVertex(chanceOfRain, 5);
+        DoubleVertex humidityObservation = new GaussianVertex(humidity, 5);
+        humidityObservation.observe(60.0);
+        chanceOfRainObservation.observe(3.0);
+
+        BayesianNetwork bayesianNetwork = new BayesianNetwork(chanceOfRainObservation.getConnectedGraph());
+
+        NetworkSamples posteriorSamples = MetropolisHastings.withDefaultConfig(random).getPosteriorSamples(
+            bayesianNetwork,
+            inputToModel,
+            2500
+        );
+
+        double averagePosteriorInput = posteriorSamples.getDoubleTensorSamples(inputToModel).getAverages().scalar();
+
+        Assert.assertEquals(29., averagePosteriorInput, 0.1);
+    }
+
+    private void modelExecution(Map<VertexLabel, Vertex<? extends Tensor>> inputs) {
+        double temperature = inputs.get(new VertexLabel("Temperature")).getValue().asFlatDoubleArray()[0];
+        try {
+            double chanceOfRain = weatherModel.blackBoxRainModel(temperature);
+            double humidity = weatherModel.blackBoxHumidityModel(temperature);
```

Comment:
Wouldn't it make more sense for the model execution to be a method on `SimpleWeatherModel`? It feels wrong that `blackBoxRainModel` and `blackBoxHumidityModel` are public. I know this is only test code but it serves an important role of demonstrating how this would work in a real scenario.

Reply:
Yeah very good point - looks much better with that logic living on the weather model. I've also moved across the logic of how to read the output once the model has executed.